pkg/controller/syncset/syncset_controller.go

```diff
@@ -0,0 +1,169 @@
+// and Start it when the Manager is Started.
+func (a *Adder) Add(mgr manager.Manager) error {
+	if !operations.HasValidationOperations() {
+		return nil
+	}
+
+	r, err := newReconciler(mgr, a.CacheManager, a.ControllerSwitch, a.Tracker)
+	if err != nil {
+		return err
+	}
+
+	return add(mgr, r)
+}
+
+func (a *Adder) InjectCacheManager(o *cm.CacheManager) {
+	a.CacheManager = o
+}
+
+func (a *Adder) InjectControllerSwitch(cs *watch.ControllerSwitch) {
+	a.ControllerSwitch = cs
+}
+
+func (a *Adder) InjectTracker(t *readiness.Tracker) {
+	a.Tracker = t
+}
+
+func newReconciler(mgr manager.Manager, cm *cm.CacheManager, cs *watch.ControllerSwitch, tracker *readiness.Tracker) (*ReconcileSyncSet, error) {
+	if cm == nil {
+		return nil, fmt.Errorf("CacheManager must be non-nil")
+	}
+	if tracker == nil {
+		return nil, fmt.Errorf("ReadyTracker must be non-nil")
+	}
+
+	return &ReconcileSyncSet{
+		reader:       mgr.GetClient(),
+		scheme:       mgr.GetScheme(),
+		cs:           cs,
+		cacheManager: cm,
+		tracker:      tracker,
+	}, nil
+}
+
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	c, err := controller.New(ctrlName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(source.Kind(mgr.GetCache(), &syncsetv1alpha1.SyncSet{}), &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileSyncSet{}
+
+// ReconcileSyncSet reconciles a SyncSet object.
+type ReconcileSyncSet struct {
+	reader client.Reader
+
+	scheme       *runtime.Scheme
+	cacheManager *cm.CacheManager
+	cs           *watch.ControllerSwitch
+	tracker      *readiness.Tracker
+}
+
+func (r *ReconcileSyncSet) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	// Short-circuit if shutting down.
+	if r.cs != nil {
+		running := r.cs.Enter()
+		defer r.cs.Exit()
+		if !running {
+			return reconcile.Result{}, nil
+		}
+	}
+
+	syncsetTr := r.tracker.For(syncsetGVK)
+	exists := true
+	syncset := &syncsetv1alpha1.SyncSet{}
+	err := r.reader.Get(ctx, request.NamespacedName, syncset)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			exists = false
+		} else {
+			// Error reading the object - requeue the request.
+			return reconcile.Result{}, err
+		}
+	}
+	sk := aggregator.Key{Source: "syncset", ID: request.NamespacedName.String()}
+
+	if !exists || !syncset.GetDeletionTimestamp().IsZero() {
+		log.V(logging.DebugLevel).Info("handling SyncSet delete", "instance", syncset)
+
+		if err := r.cacheManager.RemoveSource(ctx, sk); err != nil {
+			syncsetTr.TryCancelExpect(syncset)
+
+			return reconcile.Result{}, fmt.Errorf("synceset-controller: error removing source: %w", err)
+		}
+
+		syncsetTr.CancelExpect(syncset)
+
+		return reconcile.Result{}, nil
+	}
+
+	log.V(logging.DebugLevel).Info("handling SyncSet update", "instance", syncset)
+	gvks := []schema.GroupVersionKind{}
```

Comment:
nit: lines 155-158 feel like a nice helper function that would improve readability.

Reply:
uuu, I actually did have something for this 💯